beacon_node/network/src/sync/network_context.rs

```diff
@@ -401,69 +462,93 @@ impl<T: BeaconChainTypes> SyncNetworkContext<T> {
-
-                (
-                    Some(column_indexes.into_iter().collect::<Vec<_>>()),
-                    Some(data_column_requests),
-                )
-            } else {
-                (None, None)
-            };
+                    .collect::<Result<Vec<_>, _>>()
+            })
+            .transpose()?;
 
         let expected_blobs = blobs_req_id.is_some();
         let info = RangeBlockComponentsRequest::new(
             expected_blobs,
-            expects_columns,
+            data_column_requests.as_ref().map(|_| {
+                self.network_globals()
+                    .sampling_columns
+                    .iter()
+                    .copied()
+                    .collect()
+            }),
             data_column_requests.map(|items| items.len()),
         );
         self.components_by_range_requests.insert(id, info);
 
         Ok(id.id)
     }
 
-    fn make_columns_by_range_requests(
+    fn select_columns_by_range_peers_to_request(
         &self,
-        request: BlocksByRangeRequest,
         custody_indexes: &HashSet<ColumnIndex>,
-    ) -> Result<HashMap<PeerId, DataColumnsByRangeRequest>, RpcRequestSendError> {
-        let mut peer_id_to_request_map = HashMap::new();
+        peers: &HashSet<PeerId>,
+        active_request_count_by_peer: HashMap<PeerId, usize>,
+        peers_to_deprioritize: &HashSet<PeerId>,
+    ) -> Result<HashMap<PeerId, Vec<ColumnIndex>>, RpcRequestSendError> {
+        let mut peer_id_to_request_map = HashMap::<PeerId, Vec<ColumnIndex>>::new();
 
         for column_index in custody_indexes {
-            // TODO(das): The peer selection logic here needs to be improved - we should probably
-            // avoid retrying from failed peers, however `BatchState` currently only tracks the peer
-            // serving the blocks.
-            let Some(custody_peer) = self.get_random_custodial_peer(*column_index) else {
+            // Strictly consider peers that are custodials of this column AND are part of this
+            // syncing chain. If the forward range sync chain has few peers, it's likely that this
+            // function will not be able to find peers on our custody columns.
+            let Some(custody_peer) = peers
+                .iter()
+                .filter(|peer| {
+                    self.network_globals()
+                        .is_custody_peer_of(*column_index, peer)
+                })
+                .map(|peer| {
+                    (
+                        // If contains -> 1 (order after), not contains -> 0 (order first)
+                        peers_to_deprioritize.contains(peer),
+                        // Prefer peers with less overall requests
+                        // Also account for requests that are not yet issued tracked in peer_id_to_request_map
+                        active_request_count_by_peer.get(peer).copied().unwrap_or(0)
+                            + peer_id_to_request_map
```

Comment:
We're mixing request count + column count here, i feel like we shouldn't count the columns in the current request, because this could skew the numbers, e.g. if we have 2 active "requests" to peer_1, and now we are requesting 3 columns from peer_2 that has 0 active request, we will prioritise sending to peer_1, which actually has more active requests.